rslib/src/browser_rows.rs

```diff
@@ -0,0 +1,356 @@
+use crate::{
+    card::{Card, CardID, CardQueue, CardType},
+    collection::Collection,
+    decks::{Deck, DeckID},
+    notes::Note,
+    notetype::{CardTemplate, NoteType, NoteTypeKind},
+    scheduler::{timespan::time_span, timing::SchedTimingToday},
+    template::RenderedNode,
+    text::{extract_av_tags, html_to_text_line},
+    timestamp::{TimestampMillis, TimestampSecs},
+};
+
+const CARD_RENDER_COLUMNS: [&str; 2] = ["question", "answer"];
+
+#[derive(Debug, PartialEq)]
+pub struct Row {
+    pub cells: Vec<Cell>,
+    pub color: Color,
+    pub font: Font,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Cell {
+    pub text: String,
+    pub is_rtl: bool,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum Color {
+    Default,
+    Marked,
+    Suspended,
+    FlagRed,
+    FlagOrange,
+    FlagGreen,
+    FlagBlue,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Font {
+    pub name: String,
+    pub size: u32,
+}
+
+struct RowContext<'a> {
+    col: &'a Collection,
+    card: Card,
+    note: Note,
+    notetype: Arc<NoteType>,
+    deck: Option<Deck>,
+    original_deck: Option<Option<Deck>>,
+    i18n: &'a I18n,
+    timing: SchedTimingToday,
+    render_context: Option<RenderContext>,
+}
+
+/// The answer string needs the question string but not the other way around, so only build the
+/// answer string when needed.
+struct RenderContext {
+    question: String,
+    answer_nodes: Vec<RenderedNode>,
+}
+
+fn card_render_required(columns: &[String]) -> bool {
+    columns
+        .iter()
+        .any(|c| CARD_RENDER_COLUMNS.contains(&c.as_str()))
+}
+
+impl Collection {
+    pub fn browser_row_for_card(&mut self, id: CardID) -> Result<Row> {
+        let columns: Vec<String> = self.get_config_optional("activeCols").unwrap_or_else(|| {
+            vec![
+                "noteFld".to_string(),
+                "template".to_string(),
+                "cardDue".to_string(),
+                "deck".to_string(),
+            ]
+        });
+        let mut context = RowContext::new(self, id, card_render_required(&columns))?;
+
+        Ok(Row {
+            cells: columns
+                .iter()
+                .map(|column| context.get_cell(column))
+                .collect::<Result<_>>()?,
+            color: context.get_row_color(),
+            font: context.get_row_font()?,
+        })
+    }
+}
+
+impl RenderContext {
+    fn new(col: &mut Collection, card: &Card, note: &Note, notetype: &NoteType) -> Result<Self> {
+        let render = col.render_card(
+            note,
+            card,
+            notetype,
+            notetype.get_template(card.template_idx)?,
+            true,
+        )?;
+        let qnodes_text = render
+            .qnodes
+            .iter()
+            .map(|node| match node {
+                RenderedNode::Text { text } => text,
+                RenderedNode::Replacement {
+                    field_name: _,
+                    current_text,
+                    filters: _,
```

Comment:
Note for the future: you can do this more succinctly with Replacement { current_text, .. }

Reply:
Thanks, I thought there must be a neat way to write this. 🙂